pkg/controllers/vrep/vdbverify_reconciler.go

```diff
@@ -0,0 +1,120 @@
+/*
+(c) Copyright [2021-2024] Open Text.
+Licensed under the Apache License, Version 2.0 (the "License");
+You may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vrep
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	vapi "github.com/vertica/vertica-kubernetes/api/v1"
+	v1beta1 "github.com/vertica/vertica-kubernetes/api/v1beta1"
+	"github.com/vertica/vertica-kubernetes/pkg/controllers"
+	verrors "github.com/vertica/vertica-kubernetes/pkg/errors"
+	"github.com/vertica/vertica-kubernetes/pkg/events"
+	vmeta "github.com/vertica/vertica-kubernetes/pkg/meta"
+	"github.com/vertica/vertica-kubernetes/pkg/names"
+	"github.com/vertica/vertica-kubernetes/pkg/vk8s"
+	vrepstatus "github.com/vertica/vertica-kubernetes/pkg/vrepstatus"
+	corev1 "k8s.io/api/core/v1"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+const stateIncompatibleDB = "Incompatible"
+
+type VdbVerifyReconciler struct {
+	VRec *VerticaReplicatorReconciler
+	Vrep *v1beta1.VerticaReplicator
+	Log  logr.Logger
+}
+
+func MakeVdbVerifyReconciler(r *VerticaReplicatorReconciler, vrep *v1beta1.VerticaReplicator,
+	log logr.Logger) controllers.ReconcileActor {
+	return &VdbVerifyReconciler{
+		VRec: r,
+		Vrep: vrep,
+		Log:  log.WithName("VdbVerifyReconciler"),
+	}
+}
+
+// Reconcile will verify the verticaDBs in the vrep CR source and target exist and
+// both vertica versions support the replication feature
+func (r *VdbVerifyReconciler) Reconcile(ctx context.Context, _ *ctrl.Request) (ctrl.Result, error) {
+	// no-op if the check has already been done once
```

Comment:
```suggestion
	// no-op if the check has already been done once and was successful
```
Because the condition could be false. The question do we want to keep checking or abort the reconciliation all together.

Reply:
So do we want to keep checking until the check succeeds? @spilchen @roypaulin

Reply:
Now that I think about it, VerticaRepliactor is not something to use once and discard like scrutinize, so the current code makes sense.

Reply:
I think the CR is intended for one-time use, similar to `vrpq`, and currently it's the verification part of it (checking vdb version and deployment type) that we want to keep retrying until we succeed, and once it succeeds the actual replication process is done exactly once without retrying. After discussing with @spilchen we think this might not be the ideal behavior, as the operator would keep retrying as long as the verification fails, and it's possible that it retries while the vdb is upgrading or right after a vdb is just created and not fully ready yet. During such moments the version annotation and deployment method annotation may already indicates the new state which would satisfy the `vrep` reconciler for verification purpose but the source and target vdbs are not fully ready yet for a replication. We could in theory wait for some status conditions of source and target vdbs that indicate full readiness for replication, but that adds complexity to the implementation and may be unworthwhile. The simpler solution is to treat the whole replication attempt, including the verification part of it, as a one-shot attempt, so that it's less confusing for users. This applies similarly to the `vrpq` reconciler, and there will be a dedicated JIRA ticket for this.